### ISSUE TITLE
Google Search JavaScript Snippet not executing on navigation

### ISSUE THREAD

**Issue by rachelnabors** [defunkt/jquery-pjax#499]
On an archive page, I have this inline:

```javascript
    <script>
      (function() {
        var cx = '002428664232771263876:m1vbqio4y8u';
        var gcse = document.createElement('script');
        gcse.type = 'text/javascript';
        gcse.async = true;
        gcse.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') +
            '//www.google.com/cse/cse.js?cx=' + cx;
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(gcse, s);
      })();
    </script>
    <gcse:search></gcse:search>
```

It loads a search bar when you go to the archive's page directly via address bar, but if you navigate there via PJAX, no search bar.

**Comment by mislav** [comment]
I'm sorry to hear. I trust that you're using the latest jquery-pjax version?

Even though the address bar doesn't show up, can you check `$('script[src]').get()` before and after pjax navigation to see whether the `www.google.com/cse/cse.js` script was inserted anywhere into the DOM? Maybe the inline script executes but something else goes wrong.

**Comment by rachelnabors** [comment]
Latest version, yes!

I can confirm that `www.google.com/cse/cse.js` is inserted, as this picture shows. 
![screen shot 2015-03-13 at 4 03 05 pm](https://cloud.githubusercontent.com/assets/236306/6648739/bbb4ea8c-c99a-11e4-9c3b-d8784593818a.png)


(The nav div would show in the blue bar. See it life at http://rachelnabors.com/archive)

**Comment by mislav** [comment]
Sorry for the late reply. I'm not sure what's going on here, but because the script was successfully inserted into the page, I would guess that jQuery did its job of executing inline scripts.

When I visit the link you provided directly (no pjax navigation) I get these exceptions in Chrome:

* GET `http://www.google.com/uds/?file=ads&v=3&packages=search&async=2` net::ERR_BLOCKED_BY_CLIENT
* GET `http://www.google.com/adsense/search/async-ads.js` net::ERR_BLOCKED_BY_CLIENT

**Comment by rachelnabors** [comment]
So I've had another look at the behavior:

1) Those errors are a red herring. Other people get them on non-PJAX sites.
It's a GSE/ads thing.
2) They seem to appear only on successful displays of the search bar.
3) The search bar will display if it's the first time you've navigated to
the page, whether it's the first page you visit or you visit it via PJAX.
If you return to the page via PJAX, either by clicking the link to that
page or clicking a link to a different page then clicking the link back to
the page, it will not display. If you refresh the page manually it will.

**Comment by mislav** [comment]
When I visit this page directly in Chrome (no pjax), the search bar doesn't seem interactive. At least when I click on it, the "field" doesn't gain focus, nor does the submit button work.

This is the HTML of the search box, presumably generated by Google:

```html
  <div id="___gcse_0">
    <div class="gsc-control-cse gsc-control-cse-en">
      <div class="gsc-control-wrapper-cse" dir="ltr">
        <form accept-charset="utf-8" class="gsc-search-box gsc-search-box-tools">
          <table cellpadding="0" cellspacing="0" class="gsc-search-box">
            <tbody>
              <tr>
                <td class="gsc-input">
                  <div class="gsc-input-box" id="gsc-iw-id1">
                    <table cellpadding="0" cellspacing="0" class="gstl_50" id="gs_id50" style="width: 100%; padding: 0px;">
                      <tbody>
                        <tr>
                          <td class="gsib_a" id="gs_tti50"><input autocomplete="off" class="gsc-input" dir="ltr" id="gsc-i-id1" name="search" size="10" spellcheck="false" style="width: 100%; padding: 0px; border: none; margin: -0.0625em 0px 0px; height: 1.25em; outline: none;" title="search" type="text"></td>

                          <td class="gsib_b">
                            <div class="gsst_b" dir="ltr" id="gs_st50">
                              <a class="gsst_a" href="javascript:void(0)" style="display: none;"><span class="gscb_a" id="gs_cb50">×</span></a>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div><input id="bgresponse" name="bgresponse" type="hidden">
                </td>

                <td class="gsc-search-button"><input class="gsc-search-button gsc-search-button-v2" src="http://www.google.com/uds/css/v2/search_box_icon.png" title="search" type="image"></td>

                <td class="gsc-clear-button">
                  <div class="gsc-clear-button" title="clear results">
                    &nbsp;
                  </div>
                </td>
              </tr>
            </tbody>
          </table>

          <table cellpadding="0" cellspacing="0" class="gsc-branding">
            <tbody>
              <tr style="display: none;">
                <td class="gsc-branding-user-defined"></td>

                <td class="gsc-branding-text">
                  <div class="gsc-branding-text">
                    powered by
                  </div>
                </td>

                <td class="gsc-branding-img"><img class="gsc-branding-img" src="http://www.google.com/uds/css/small-logo.png"></td>
              </tr>
            </tbody>
          </table>
        </form>

        <div class="gsc-results-wrapper-overlay">
          <div class="gsc-results-close-btn" tabindex="0"></div>

          <div class="gsc-tabsAreaInvisible">
            <div class="gsc-tabHeader gsc-inline-block gsc-tabhActive">
              Custom Search
            </div><span class="gs-spacer"></span>
          </div>

          <div class="gsc-tabsAreaInvisible"></div>

          <div class="gsc-above-wrapper-area-invisible">
            <table cellpadding="0" cellspacing="0" class="gsc-above-wrapper-area-container">
              <tbody>
                <tr>
                  <td class="gsc-result-info-container">
                    <div class="gsc-result-info-invisible"></div>
                  </td>

                  <td class="gsc-orderby-container">
                    <div class="gsc-orderby-invisible">
                      <div class="gsc-orderby-label gsc-inline-block">
                        Sort by:
                      </div>

                      <div class="gsc-option-menu-container gsc-inline-block">
                        <div class="gsc-selected-option-container gsc-inline-block">
                          <div class="gsc-selected-option">
                            Relevance
                          </div>

                          <div class="gsc-option-selector"></div>
                        </div>

                        <div class="gsc-option-menu-invisible">
                          <div class="gsc-option-menu-item gsc-option-menu-item-highlighted">
                            <div class="gsc-option">
                              Relevance
                            </div>
                          </div>

                          <div class="gsc-option-menu-item">
                            <div class="gsc-option">
                              Date
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>

          <div class="gsc-adBlockInvisible"></div>

          <div class="gsc-wrapper">
            <div class="gsc-adBlockInvisible"></div>

            <div class="gsc-resultsbox-invisible">
              <div class="gsc-resultsRoot gsc-tabData gsc-tabdActive">
                <table cellpadding="0" cellspacing="0" class="gsc-resultsHeader">
                  <tbody>
                    <tr>
                      <td class="gsc-twiddleRegionCell">
                        <div class="gsc-twiddle">
                          <div class="gsc-title">
                            Web
                          </div>
                        </div>

                        <div class="gsc-stats"></div>

                        <div class="gsc-results-selector gsc-all-results-active">
                          <div class="gsc-result-selector gsc-one-result" title="show one result">
                            &nbsp;
                          </div>

                          <div class="gsc-result-selector gsc-more-results" title="show more results">
                            &nbsp;
                          </div>

                          <div class="gsc-result-selector gsc-all-results" title="show all results">
                            &nbsp;
                          </div>
                        </div>
                      </td>

                      <td class="gsc-configLabelCell"></td>
                    </tr>
                  </tbody>
                </table>

                <div>
                  <div class="gsc-expansionArea"></div>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="gsc-modal-background-image" tabindex="0"></div>
      </div>
    </div>
  </div>
```

It's horrid, and on top of all it doesn't even work. Isn't there an easier way to make a search box?

I know I'm avoiding looking into the supposed pjax problem, but a lot of script-related problems are solved by avoiding needing a script in the first place. The best JS sites are the ones that avoid JS as much as possible.

**Comment by rachelnabors** [comment]
I use Jekyll to make this site. As a flat site, there is no way to build a
search box of my own. Google's boxes are used by many sites, and thus, as
with PJAX, I chose to trust the solution used by the most people as I could
expect the most support on sites like Stack Overflow and here on GitHub.

I'm sorry the generated code isn't pretty enough for you. But I'd rather
have a search box with ugly markup than none at all.

Also, it works for me when I remove PJAX on my end.

**Comment by mislav** [comment]
> I'm sorry the generated code isn't pretty enough for you.

I wasn't complaining about the generated code being pretty. I was wondering why even after all this generated markup, the search bar wasn't interactive when loading the page directly. My mouse clicks on it were ignored. After opening web inspector right now, I've realized that the search box was trumped over by the heading element which had a negative top margin and a large padding:

![bloggage](https://cloud.githubusercontent.com/assets/887/6963544/de024c14-d940-11e4-9c58-a159bf072c10.png)

After removing the margin and padding of the heading element, I was finally able to verify that the search box works, although the padding of the search input element seems off, and line heights of the search results seem messed up (but that is a cascadding CSS problem that you'll probably tackle independently).

* * *

Next, I was able to reproduce the search box not working after pjax navigation. This is a problem of Google's scripts only replacing `<gcse:search>` elements once after script load, but never again afterwards, even if new `<gcse:search>` elements are added via pjax. The solution is to manually call `google.search.cse.element.go()` after each navigation, i.e. on "pjax:end".

This code example demonstrates that and also does a quick hack to fix the old search boxes restored from pjax cache after popstate navigation (such as using the Back button in the browser).

```js
$(document).on('pjax:end', function(event) {
  $('div[id^="___gcse_"]').replaceWith("<gcse:search>")
  $('table[class^="gstl_"]').remove()
  try {
    google.search.cse.element.go()
  } catch(e) { }
})
```

* * *

> As a flat site, there is no way to build a search box of my own.

As an alternative to everything above, as well as Google Custom Search, you could try:

```html
<form action="https://www.google.com/search">
  <input type=hidden name="as_sitesearch" value="rachelnabors.com">
  <input type=search name="q" placeholder="Search this site">
  <input type=submit value="Search">
</form>
```

The results don't show in a fancy popup on the same page, but maybe that's for the better. My recommendation would be to use a static HTML solution rather than loading Google's scripts, simply because in general, a [site with less JS is a happier site](https://twitter.com/mislav/statuses/558762669893623808).

**Comment by rachelnabors** [comment]
Well, it's good to know how to do that. I'll take your advice and use the
JS-free form. I wasn't aware you could do that with Google.

**Comment by mislav** [comment]
@rachelnabors I'm going to close this because we found the cause and at least 2 workarounds. It's nothing that we can "fix" in core pjax, though. Thanks for your patience.
